// bsmp_main/store/getters.js
export default {
  /**
   * 登录用户
   * @param state
   * @returns {*}
   */
  loginUser: (state) => {
    if (state && state.auth && state.auth.user) {
      return state.auth.user;
    }
    return null;
  },

  /**
   * 加载菜单列表
   * @param state
   * @returns {*}
   */
  menus: (state) => {
    if (state && state.menu && state.menu.menus) {
      return state.menu.menus;
    }
    return null;
  },

  /**
   * 第一个菜单路径
   * @param state
   * @returns {*}
   */
  firstPath: (state) => {
    return state.menu.firstPath;
  },

  /**
   * 加载菜单列表
   * @param state
   * @returns {*}
   */
  modules: (state) => {
    if (state && state.menu && state.menu.modules) {
      return state.menu.modules;
    }
    return null;
  },

  /**
   * 加载公司列表
   * @param state
   * @returns {*}
   */
  companies: (state) => {
    if (state && state.company && state.company.companies) {
      return state.company.companies;
    }
    return null;
  },

  /**
   * 当前切换公司
   * @param state
   * @returns {*}
   */
  company: (state) => {
    if (state && state.company && state.company.company) {
      return state.company.company;
    }
    return null;
  }
};
