// bsmp_main/api/index.js
import axios from 'axios';
import qs from 'qs';

const path = '/api/sys_config/';

export default {

  /**
   * 记录日志
   * @param params
   * @returns {AxiosPromise}
   */
  log (params) {
    return axios({
      method: 'post',
      url: `${path}log`,
      data: qs.stringify(params)
    });
  },

  /**
   * 登录
   * @param name
   * @param password
   * @returns {AxiosPromise}
   */
  login ({name, password}) {
    return axios({
      method: 'post',
      url: `${path}user/login`,
      data: qs.stringify({
        loginname: name.trim(),
        pwd: password.trim()
      })
    });
  },

  /**
   * 登出
   * @param token
   * @returns {AxiosPromise}
   */
  logout ({token}) {
    return axios({
      method: 'post',
      url: `${path}user/logout`,
      data: qs.stringify({
        token
      })
    });
  },

  /**
   * 修改密码
   * @param params
   * @returns {AxiosPromise}
   */
  modifyPwd (params) {
    return axios({
      method: 'post',
      url: `${path}user/modifyPwd`,
      data: qs.stringify(params)
    });
  },

  /**
   * 获取菜单列表数据
   * @param params
   * @returns {AxiosPromise}
   */
  getMenus (params) {
    return axios({
      method: 'get',
      url: `${path}menu`,
      params: params
    });
  },

  /**
   * 获取功能操作列表
   * @param params
   * @returns {AxiosPromise}
   */
  getFuncs (params) {
    return axios({
      method: 'get',
      url: `${path}menu/getAuthFun`,
      params: params
    });
  },

  /**
   * 公司列表
   * @returns {AxiosPromise}
   */
  getCompanies () {
    return axios({
      method: 'get',
      url: `${path}company/getList`
    });
  },

  /**
   * 获取字典
   * @param code
   * @returns {AxiosPromise}
   */
  getDictByPCode (code) {
    return axios({method: 'get', url: `${path}dict/findListByPCode/${code}`});
  }
};
