// sys_config/api/index.js
import axios from 'axios';
import qs from 'qs';

const path = '/api/sys_config/';

export default {

  /**
   * 公司
   * @param params
   * @returns {*}
   */
  getCompanyList (params) {
    return axios({method: 'get', url: `${path}company`, params: params});
  },
  addCompany (params) {
    return axios({method: 'post', url: `${path}company`, data: qs.stringify(params)});
  },
  editCompany (params) {
    return axios({method: 'put', url: `${path}company/${params.id}`, data: qs.stringify(params)});
  },
  removeCompany (id) {
    return axios({method: 'delete', url: `${path}company/${id}`});
  },
  removeCompanies (ids) {
    return axios({method: 'delete', url: `${path}company/deleteByIds/${ids}`});
  },
  /**
   * 角色
   * @param params
   * @returns {*}
   */
  getRoleList (params) {
    return axios({method: 'get', url: `${path}role`, params: params});
  },
  addRole (params) {
    return axios({method: 'post', url: `${path}role`, data: qs.stringify(params)});
  },
  editRole (params) {
    return axios({method: 'put', url: `${path}role/${params.id}`, data: qs.stringify(params)});
  },
  removeRole (id) {
    return axios({method: 'delete', url: `${path}role/${id}`});
  },
  removeRoles (ids) {
    return axios({method: 'delete', url: `${path}role/deleteByIds/${ids}`});
  },
  /**
   * 用户
   * @param params
   * @returns {*}
   */
  getUserList (params) {
    return axios({method: 'get', url: `${path}user`, params: params});
  },
  addUser (params) {
    return axios({method: 'post', url: `${path}user`, data: qs.stringify(params)});
  },
  editUser (params) {
    return axios({method: 'put', url: `${path}user/${params.id}`, data: qs.stringify(params)});
  },
  removeUser (id) {
    return axios({method: 'delete', url: `${path}user/${id}`});
  },
  removeUsers (id) {
    return axios({method: 'delete', url: `${path}user/deleteByIds/${id}`});
  },
  resetPwd (params) {
    return axios({method: 'post', url: `${path}user/resetPwd`, data: qs.stringify(params)});
  },
  /**
   * 菜单信息
   * @param params
   * @returns {*}
   */
  getMenuList (params) {
    return axios({method: 'get', url: `${path}menu`, params: params});
  },
  addMenu (params) {
    return axios({method: 'post', url: `${path}menu`, data: qs.stringify(params)});
  },
  editMenu (params) {
    return axios({method: 'put', url: `${path}menu/${params.id}`, data: qs.stringify(params)});
  },
  removeMenu (id) {
    return axios({method: 'delete', url: `${path}menu/${id}`});
  },
  authRoleMenus (params) {
    return axios({method: 'post', url: `${path}menu/setAuth`, data: qs.stringify(params)});
  },

  getUserMenuList (params) {
    return axios({method: 'get', url: `${path}menu/userMenu`, params: params});
  },
  authUserMenus (params) {
    return axios({method: 'post', url: `${path}menu/setUserAuth`, data: qs.stringify(params)});
  },

  /**
   * 字典信息
   * @param params
   * @returns {*}
   */
  getDictList (params) {
    return axios({method: 'get', url: `${path}dict`, params: params});
  },
  getDictByPCode (code) {
    return axios({method: 'get', url: `${path}dict/findListByPCode/${code}`});
  },
  addDict (params) {
    return axios({method: 'post', url: `${path}dict`, data: qs.stringify(params)});
  },
  editDict (params) {
    return axios({method: 'put', url: `${path}dict/${params.id}`, data: qs.stringify(params)});
  },
  removeDict (id) {
    return axios({method: 'delete', url: `${path}dict/${id}`});
  },
  /**
   * 操作日志信息
   * @param params
   * @returns {*}
   */
  getLogList (params) {
    return axios({method: 'get', url: `${path}log`, params: params});
  }
};
